src/volume/streamed.rs

```diff
@@ -0,0 +1,449 @@
+        let buf = Vec::with_capacity(6);
+        let buf = {
+            let slice = volume
+                .next_inline(buf)
+                .expect("1st slice should exist")
+                .expect("should not fail to construct the volume");
+
+            assert_eq!(slice.dim(), &[2, 3]);
+            assert_eq!(slice.data_type(), NiftiType::Uint8);
+            assert_eq!(slice.raw_data(), &[1, 3, 5, 7, 9, 11]);
+            slice.into_raw_data()
+        };
+        {
+            let slice = volume
+                .next_inline(buf)
+                .expect("2nd slice should exist")
+                .expect("should not fail to construct the volume");
+
+            assert_eq!(slice.dim(), &[2, 3]);
+            assert_eq!(slice.data_type(), NiftiType::Uint8);
+            assert_eq!(slice.raw_data(), &[13, 15, 17, 19, 21, 23]);
+        }
+        assert!(volume.next().is_none());
+    }
+
+    #[test]
+    fn test_streamed_ranked() {
+        let volume_data = &[1, 3, 5, 7, 9, 11, 13, 15, 17, 19, 21, 23];
+        let header = NiftiHeader {
+            dim: [4, 2, 3, 2, 1, 0, 0, 0],
+            datatype: NiftiType::Uint8 as i16,
+            scl_slope: 1.,
+            scl_inter: 0.,
+            endianness: Endianness::native(),
+            ..NiftiHeader::default()
+        };
+
+        let mut volume =
+            StreamedNiftiVolume::from_reader_rank(&volume_data[..], &header, 2).unwrap();
+
+        assert_eq!(volume.dim(), &[2, 3, 2, 1]);
+        assert_eq!(volume.slice_dim(), &[2, 3]);
+        assert_eq!(volume.slices_read(), 0);
+
+        {
+            let slice = volume
+                .next()
+                .expect("1st slice should exist")
+                .expect("should not fail to construct the volume");
+
+            assert_eq!(slice.dim(), &[2, 3]);
+            assert_eq!(slice.data_type(), NiftiType::Uint8);
+            assert_eq!(slice.raw_data(), &[1, 3, 5, 7, 9, 11]);
+        }
+        {
+            let slice = volume
+                .next()
+                .expect("2nd slice should exist")
+                .expect("should not fail to construct the volume");
+
+            assert_eq!(slice.dim(), &[2, 3]);
+            assert_eq!(slice.data_type(), NiftiType::Uint8);
+            assert_eq!(slice.raw_data(), &[13, 15, 17, 19, 21, 23]);
+        }
+        assert!(volume.next().is_none());
+    }
+
+    #[test]
+    fn test_streamed_lesser_rank() {
+        let volume_data = &[
+            1, 0, 3, 0, 5, 0, 7, 0, 9, 0, 11, 0, 13, 0, 15, 0, 17, 0, 19, 0, 21, 0, 23, 0,
+        ];
+        let header = NiftiHeader {
+            dim: [3, 2, 3, 2, 0, 0, 0, 0],
+            datatype: NiftiType::Uint16 as i16,
+            scl_slope: 1.,
+            scl_inter: 0.,
+            endianness: Endianness::Little,
+            ..NiftiHeader::default()
+        };
+
+        let mut volume =
+            StreamedNiftiVolume::from_reader_rank(&volume_data[..], &header, 1).unwrap();
```

Comment:
I really don't understand why you need to create a slice out of `volume_data`, which is already a slice. Ah, hum, it works if you write

    let volume_data: &[u8] = &[

I thought that the right type would be automatically known by the compiler.

Reply:
This is one of the cases where the compiler doesn't perform deref coercion. The input for `from_reader_rank` is something that implements `Read`, but references to arrays don't implement that and the compiler won't turn it into a slice automatically.